xarray_filters/mldataset.py

```diff
@@ -26,8 +26,44 @@ class MLDataset(xr.Dataset):
     class defintion - wrap the documentation for to_features,
     from_features, and concat_ml_features, and chain methods
     '''
-
-    def to_features(self,*args, **kwargs):
+    def load(self, filepaths, meta=None, verify=True, loader=None):
+        """Register list of filepaths (which may include directory paths) as
+        data sources. This method may be called multiple times to register
+        different file/directory paths separately. Each data source must be
+        loadable via xarray or dask; complete list of supported data sources
+        available here:
+
+        Xarray data sources: http://xarray.pydata.org/en/stable/io.html
+        Dask data sources: https://dask.pydata.org/en/latest/dataframe-api.html#create-dataframes
+
+        Some lightweight verification steps may be performed by this method to
+        confirm that the data sources are valid/reachable, but data source(s)
+        will only be loaded on-demand (as needed) to conserve compute resources.
+
+        Parameters:
+            :filepaths: list of str/bytes; e.g. file paths (HDF4 / 5 or NetCDF) or directory paths (TIF)
+            :meta:      dict of str->dict; meta data for "filepaths". Each (key, value) pair represents a filepath and its meta data.
+            :verify:    bool; False means "skip verification steps".
+            :loader:    function or None; function can be `xr.open_rasterio`, for example. If None, uses `earthio.load_layers`.
+        """
+        if not hasattr(self, '__data_sources'):
+            self.__data_sources = {}
+        for fpath_or_dpath in filepaths:
+            info = {
+                'registered': False,
+                'meta': meta.get(fpath_or_dpath),
+                'loader': loader,
+            }
+            if verify:
+                # Verification steps go here
+                pass
+            info['registered'] = True
+            if info.get('loader', None) is None:
+                import earthio
+                info['loader'] = earthio.load_layers
```

Comment:
Does storing a function handle here in `info` (later `self`) affect serialization options (`dill` is used currently)?  Elsewhere in Elm (`EaSearchCV` methods) `_open` and `_close` are used for before/after fitting to delete attributes that are not-serializable (dynamically defined data structures from `deap` there), so we can handle it that way here as well, if needed.

Reply:
[Based on this](https://medium.com/@emlynoregan/serialising-all-the-functions-in-python-cd880a63b591) I think we should be OK as long as we only use functions and not methods from classes, although I could be wrong. If you think it would be better I can switch to storing a string, and have a test that does a round-trip for the `earthio.load_layers()` case.

Reply:
@gbrener  After thinking about it more, it seems like `MLDataset.load` should just call `super(MLDataset, self).load()` to lazy load already defined arrays.  I don't think it should relate to filenames, etc.  I was confusing `open_dataset` with `Dataset.load`. That's a quick fix (the `super` thing), then we just have a documentation burden on how to explain `load_layers` vs `open_dataset` vs `open_mfdataset` and how `MLDataset.load`/`Dataset.load` are typically not called but can be used to convert from lazy to eager.  

I took some notes that could be the start of documentation on file loading choices (a bit of thread drift here into documentation needs):

I have one file, so I use `load_layers` for one file:
 * NetCDF
 * HDF4 / HDF5
 * GeoTiff
 * Grib 

Questions:
 * What if my "one file" for a NetCDF is actually a URL for an OpenDAP endpoint?  Is `load_layers` broken in this case?  We should have this working as in `xarray.open_dataset` with a NetCDF URL.

I have about 8 or 15 GeoTiffs where each is a separate satellite band:
 * Pass the directory of them as "filename" to `load_layers`, use `LayerSpec` to define which bands you want in that directory

I have many files that can be loaded with xarray.open_mfdataset:
 * Just use `xarray.open_mfdataset`.  This supports NetCDF, Grib.  Not sure about HDF5?  Then call `MLDataset(dset)`

Here's the signature for `open_dataset`.  Maybe we want to walk through how some of the arguments compare to arguments of `load_layers`
```
open_dataset(filename_or_obj, group=None, decode_cf=True, mask_and_scale=True, decode_times=True, autoclose=False, concat_characters=True, decode_coords=True, engine=None, chunks=None, lock=None, cache=None, drop_variables=None)
```
    
filename_or_obj: Same in `load_layers` but it may be a directory in the case of GeoTiffs for `load_layers`

group : This `group` argument for `open_dataset` is called `layer_spec` in `load_layers`.  A list of `LayerSpec` objects can control which groups are loaded and the `load_layers` style of `LayerSpec` applies to all the file types, not just NetCDF.

decode_cf : We should make sure `load_layers` decodes (True) according to CF conventions, equivalent to passing `decode_cf=True`.

mask_and_scale : This would be nice to support, but not critical.  Here's the help:
```
If True, replace array values equal to _FillValue with NA and scale values according to the formula original_values * scale_factor + add_offset, where _FillValue, scale_factor and add_offset are taken from variable attributes (if they exist). If the _FillValue or missing_value attribute contains multiple values a warning will be issued and all array values matching one of the multiple values will be replaced by NA.
```

Reply:
Moving this into a separate issue: https://github.com/ContinuumIO/earthio/issues/40

I figure we might be able to resume discussion there. Should I simply close this PR? I'm not sure if it's a net benefit to merge or not at this point.